libs/full/async_cuda/src/cuda_event_callback.cpp

```diff
@@ -0,0 +1,231 @@
+        event_callback_function_type f;
+    };
+
+    using event_callback_queue_type =
+        concurrency::ConcurrentQueue<event_callback>;
+    using event_callback_vector_type = std::vector<event_callback>;
+
+    event_callback_vector_type& get_event_callback_vector()
+    {
+        static event_callback_vector_type event_callback_vector;
+        return event_callback_vector;
+    }
+
+    event_callback_queue_type& get_event_callback_queue()
+    {
+        static event_callback_queue_type event_callback_queue;
+        return event_callback_queue;
+    }
+
+    void add_to_event_callback_queue(event_callback&& continuation)
+    {
+        HPX_ASSERT_MSG(get_register_polling_count() != 0,
+            "CUDA event polling has not been enabled on any pool. Make sure "
+            "that CUDA event polling is enabled on at least one thread pool.");
+
+        get_event_callback_queue().enqueue(std::move(continuation));
+
+        cud_debug.debug(debug::str<>("event queued"), "event",
+            debug::hex<8>(continuation.event), "callbacks in vector",
+            debug::dec<3>(get_event_callback_vector().size()));
+    }
+
+    void add_to_event_callback_vector(event_callback&& continuation)
+    {
+        get_event_callback_vector().push_back(std::move(continuation));
+
+        cud_debug.debug(
+            debug::str<>("event callback moved from queue to vector"), "event",
+            debug::hex<8>(continuation.event), "callbacks in vector",
+            debug::dec<3>(get_event_callback_vector().size()));
+    }
+
+    void add_event_callback(
+        event_callback_function_type&& f, cudaStream_t stream)
+    {
+        cudaEvent_t event;
+        if (!cuda_event_pool::get_event_pool().pop(event))
+        {
+            HPX_THROW_EXCEPTION(
+                invalid_status, "add_event_callback", "could not get an event");
+        }
+        check_cuda_error(cudaEventRecord(event, stream));
+
+        detail::add_to_event_callback_queue(
+            event_callback{event, std::move(f)});
+    }
+
+    // Background progress function for async CUDA operations. Checks for completed
+    // cudaEvent_t and calls the associated callback when ready. We first process
+    // events that have been added to the vector of events, which should be
+    // processed under a lock.  After that we process events that have been added to
+    // the lockfree queue. If an event from the queue is not ready it is added to
+    // the vector of events for later checking.
+    hpx::threads::policies::detail::polling_status poll()
+    {
+        using hpx::threads::policies::detail::polling_status;
+
+        auto& event_callback_vector = detail::get_event_callback_vector();
+
+        // Don't poll if another thread is already polling
+        std::unique_lock<hpx::cuda::experimental::detail::mutex_type> lk(
+            detail::get_vector_mtx(), std::try_to_lock);
+        if (!lk.owns_lock())
+        {
+            if (cud_debug.is_enabled())
```

Comment:
can it be turn to a if constexpr? not sure

Reply:
Could with C++17, good point! As far as I understood @biddisco verified that this gets optimized out in any case, but constexpr if would guarantee it.